### PR TITLE
[shape_poly] Add support for jnp.cum{sum,prod,max,min} with shape polymorphism

### PR DESCRIPTION
**Pull request by gnecula**
Unfortunately, on CPU and GPU where we use associative scan, we cannot support shape polymorphism with native lowering, when the axis over which we reduce is non-constant size. Improve the error in that case.

### CODE REVIEW

**Comment by gnecula** [comment]
@rxsang @maxwillzq PTAL

File: jax/experimental/jax2tf/tests/shape_poly_test.py [hunk 1]
```diff
@@ -1855,6 +1855,19 @@ def f_jax(x):
                 lambda x: lax_control_flow.cummax(x, axis=1, reverse=False),
                 arg_descriptors=[RandArg((3, 4, 5), _f32)],
                 poly_axes=[0]),
+    PolyHarness("jnp.cumsum", "reduce_axis=poly",
+                lambda x: jnp.cumsum(x, axis=0),
+                arg_descriptors=[RandArg((3, 5), _f32)],
+                poly_axes=[0],
+                expect_error=(
+                  (None, None) if (not config.jax2tf_default_experimental_native_lowering or
```
Comment (maxwillzq) [inline jax/experimental/jax2tf/tests/shape_poly_test.py:1863]: Question:   I do not understand why need check "native_lowering" here.  Since the error message on loops.py only tell users "avoid to run it on TPU".

File: jax/experimental/jax2tf/tests/shape_poly_test.py [hunk 2]
```diff
@@ -2567,6 +2580,10 @@ def test_harness(self, harness: PolyHarness):
         raise unittest.SkipTest(
             "native lowering with shape polymorphism not implemented for JAX primitives still using HLO fallback lowering; b/261682623")
 
+      if harness.fullname == "jnp.cumsum_reduce_axis=poly" and jtu.device_under_test() == "tpu":
+        # https://github.com/openxla/stablehlo/issues/1258
+        raise unittest.SkipTest(
+            "native lowering with shape polymorphism not implemented for window_reductions on TPU")
```
Comment (maxwillzq) [inline jax/experimental/jax2tf/tests/shape_poly_test.py:2586]: same question as before.  Since shape_poly_test.py can run both non_native_lowering and native_lowering.   so Non-native-lowering with TPU will also skip this unit test. 
Do we need mention "native_lowering" in the error message?

File: jax/experimental/jax2tf/tests/shape_poly_test.py (hunk 1 above)
Comment (gnecula) [reply jax/experimental/jax2tf/tests/shape_poly_test.py:1863]: We need the check here because we want to capture the precise condition under which we get an error: native lowering and device != "TPU". Note that if I change the condition the test will fail, because an expected error is not appearing.

The error message could be more precise: you get this error if you try to use associative scan with shape polymorphism. For non-native lowering, jax2tf does not use associative scan (and jax2tf is independent of the platform). However, there is the flag `--jax2tf_associative_scan_reductions` that makes jax2tf non-native use associative scan.

With native lowering you get platform-dependent behavior, just like JAX native: will use associative scan for CPU and GPU.

File: jax/experimental/jax2tf/tests/shape_poly_test.py (hunk 2 above)
Comment (gnecula) [reply jax/experimental/jax2tf/tests/shape_poly_test.py:2586]: This whole conditional is under a bigger conditional that only applies on native lowering.

**Review by maxwillzq — APPROVED** [review]

**Review by rxsang — APPROVED** [review]
